source/recipes/faq.md

```diff
@@ -62,16 +62,6 @@ See <https://nixos.wiki/wiki/Nix_channels>
 
 See <https://github.com/nix-community/home-manager>
 
-### Are there some known impurities in builds?
-
-Yes.
-
-- CPU (we try hard to avoid compiling native instructions, but rather hardcode supported ones)
-- current time/date
-- FileSystem (ext4 has a known bug creating [empty files on power loss](https://github.com/NixOS/nixpkgs/issues/15581))
```

Comment:
This should also be in the updated list

Reply:
Digging further on the issue there's this:

> [domenkozar](https://github.com/domenkozar) commented [May 23, 2016](https://github.com/NixOS/nixpkgs/issues/15581#issuecomment-220964000)
> It turns out the issue was completely different impurity: https://github.com/snabblab/snabblab-nixos/issues/42#issuecomment-220933393
> 
> I'll open a trivial PR soon to address this (testing now if it works)

And:

> [domenkozar](https://github.com/domenkozar) commented [May 23, 2016](https://github.com/snabblab/snabblab-nixos/issues/42#issuecomment-220938218)
> Confirmed! It's due to https://github.com/snabblab/snabblab-nixos/blob/master/modules/sudo-in-builds.nix
> 
> If I uncomment that module, initrd builds successfully.

Reply:
The filesystem can definitely be an impurity, see e.g. https://github.com/NixOS/nixpkgs/pull/104457#issuecomment-989106739

Reply:
@infinisil, do you think "File system dependency injections" would be a good way to generalize this case?

https://blog.ploeh.dk/2017/01/27/dependency-injection-is-passing-an-argument/
https://en.wikipedia.org/wiki/Dependency_injection

And thank you very much for your feedback (and patience). I lack the intuition/knowledge but am learning a great deal!

Reply:
No problem :)

I don't think dependency injection would be appropriate here. That term is mainly for testing when you want to change some underlying dependency with something that you can control in the test. But we're not changing any dependency here, it's just that the underlying file system differs from system to system.

Reply:
Here's a suggested wording:

```markdown
- The filesystem used for building (see also [`TMPDIR`](https://nixos.org/manual/nix/stable/command-ref/env-common.html?highlight=TMPDIR#env-TMPDIR)).
```